### ISSUE TITLE
Bug: "no-unused-vars" incorrectly flags object destructuring

### ISSUE THREAD

**Issue by zachsa** [eslint/eslint#15136]
### Environment

Node version: 16.10
npm version: 7
Local ESLint version: 7.32
Global ESLint version: 7.32
Operating System: Linux Ubuntu (WSL 2)

### What parser are you using?

@babel/eslint-parser

### I like to use object destructing as a means of filtering away unused fields. For example:

```js
const {fieldA, ...otherFields} = obj
console.log(otherFields) // Doesn't include fieldA
```

I find this a very convenient means of blacklisting fields. The `fieldA` variable is definitely used in this case, since NOT including results in that field being included in the `otherFields` destructured object. When I do this I always have to disable the rule to avoid linting errors:

```js
// eslint-disable-next-line no-unused-vars
const {fieldA, ...otherFields} = obj
console.log(otherFields) // Doesn't include fieldA
```

### What did you expect to happen?

In this case - `fieldA` should not be identified as an unused variable.

### What actually happened?

`fieldA` is identified as an unused variable, even though the declaration of this variable in and of itself changes program functionality. It is a 'used' variable in this sense.

### Participation

- [X] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
Hi @zachsa, thanks for the issue!

> ```js
> const {fieldA, ...otherFields} = obj
> console.log(otherFields) // Doesn't include fieldA
> ```

This rule has [`ignoreRestSiblings`](https://eslint.org/docs/rules/no-unused-vars#ignorerestsiblings) option. When the option is set to `true`, `fieldA` will be allowed, as in [this demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLXZhcnM6IFtcImVycm9yXCIsIHsgXCJpZ25vcmVSZXN0U2libGluZ3NcIjogdHJ1ZSB9XSovXG5cbmNvbnN0IHtmaWVsZEEsIC4uLm90aGVyRmllbGRzfSA9IG9ialxuY29uc29sZS5sb2cob3RoZXJGaWVsZHMpIC8vIERvZXNuJ3QgaW5jbHVkZSBmaWVsZEFcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=). Can you check your configuration for `no-unused-vars`? Default value for this option is `false`, so if the option isn't set then this is correct behavior,

**Comment by zachsa** [comment]
Thank you!

I did not know how to search for that. I find that this can be added to the config like so:

```
 ...,
  rules: {
    'no-prototype-builtins': 0,
    ...,
    'no-unused-vars': ['error', { vars: 'all', args: 'after-used', ignoreRestSiblings: true }],
  },
 ...
```
